sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,239 @@
+        private readonly IOperation<TResult, TResponseType> _operation;
+
+        public OperationInternal(string operationTypeName,
+            ClientDiagnostics clientDiagnostics,
+            IOperation<TResult, TResponseType> operation,
+            TimeSpan? defaultPollingInterval = null) : base(operationTypeName, clientDiagnostics, defaultPollingInterval)
+        {
+            _operation = operation;
+        }
+
+        protected override async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = ClientDiagnostics.CreateScope($"{OperationTypeName}.UpdateStatus");
+
+            _operation.AddAttributes(scope);
+            scope.Start();
+
+            try
+            {
+                Response<TResponseType> response = async
+                    ? await _operation.GetResponseAsync(cancellationToken).ConfigureAwait(false)
+                    : _operation.GetResponse(cancellationToken);
+
+                Response rawResponse = response.GetRawResponse();
+                RawResponse = rawResponse;
+
+                var status = _operation.UpdateState(response);
+
+                if (status == CompletionStatus.Succeeded)
+                {
+                    Value = _operation.ParseResponse(response);
+                    HasCompleted = true;
+                }
+                else if (status == CompletionStatus.Failed)
+                {
+                    OperationFailedException = _operation.GetOperationFailedException(response);
+                    HasCompleted = true;
+
+                    throw OperationFailedException;
+                }
+
+                return rawResponse;
+            }
+            catch (RequestFailedException e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+            catch (Exception e)
+            {
+                var requestFailedException = new RequestFailedException("Unexpected failure.", e);
+                scope.Failed(requestFailedException);
+                throw requestFailedException;
+            }
+        }
+    }
+
+    internal interface IOperation<TResult>
+    {
+        Task<Response> GetResponseAsync(CancellationToken cancellationToken);
+
+        Response GetResponse(CancellationToken cancellationToken);
+
+        CompletionStatus UpdateState(Response response);
+
+        TResult ParseResponse(Response response);
+
+        RequestFailedException GetOperationFailedException(Response response);
+
+        void AddAttributes(DiagnosticScope scope);
+    }
+
+    internal interface IOperation<TResult, TResponseType>
+    {
+        Task<Response<TResponseType>> GetResponseAsync(CancellationToken cancellationToken);
+
+        Response<TResponseType> GetResponse(CancellationToken cancellationToken);
+
+        CompletionStatus UpdateState(Response<TResponseType> response);
+
+        TResult ParseResponse(Response<TResponseType> response);
```

Comment:
`ParseResponse` is only used to obtain `Value` from the response once the operation successfully completes. For simplicity, I'm considering removing it and expecting developers to set `Value` manually in `UpdateState`. We use `UpdateState` to update operation-specific properties as well (`StartDate`, `EndDate`, etc).